[V1] Set structured output backend to `auto` by default

russellb opened the pull request:
The previous default was `xgrammar` and users had to opt-in to fallback
behavior. After more thought, `auto` seems like a better default as it
lets us do our best to satisfy all requests. Users can still pin vllm to
a single backend if desired.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

russellb committed:
[V1] Set structured output backend to `auto` by default

The previous default was `xgrammar` and users had to opt-in to fallback
behavior. After more thought, `auto` seems like a better default as it
lets us do our best to satisfy all requests. Users can still pin vllm to
a single backend if desired.

Make `auto` work for V0 in case it gets specified there, as well.

Signed-off-by: Russell Bryant <rbryant@redhat.com>

njhill commented on vllm/config.py:
```diff
@@ -2855,10 +2855,10 @@ def __post_init__(self):
-            valid_guided_backends = v1_valid_guided_backends
-        else:
+        if envs.VLLM_USE_V1 == "0":
             valid_guided_backends = v0_valid_guided_backends
+        else:
+            valid_guided_backends = v1_valid_guided_backends
```
What's the reason for this change? `envs.VLLM_USE_V1` is a `bool`...

russellb replied:
sigh, thanks for pointing that out

russellb replied:
i misinterpreted a failure and made a change too fast trying to wrap up my day - I'm going to revert this, though there may still be a problem...

njhill commented on vllm/engine/arg_utils.py:
```diff
@@ -391,7 +391,7 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
             '--guided-decoding-backend',
             type=str,
-            default='xgrammar',
+            default="xgrammar" if envs.VLLM_USE_V1 == "0" else "auto",
```
```suggestion
            default="xgrammar" if not envs.VLLM_USE_V1 else "auto",
```

hmellor reviewed (changes requested):
This would only change the default behaviour if vLLM is started via CLI. The default for `LLM` would still be `xgrammar`.

This is a bit of a general gotcha with the way we define defaults. They can either be defined in:

- `SomethingConfig` - as a default in the dataclass
- `EngineArgs` - as a default in this dataclass
- `EngineArgs.add_cli_args` - as a default value for an `argparse` argument

It's not something for this PR, but we should probably decide where all defaults should live and do a little refactor.

russellb commented:
@hmellor Thank you! Great catch. 100% agree with the need to refactor this ... very confusing

hmellor commented on vllm/config.py:
```diff
@@ -2886,9 +2886,11 @@ def get_served_model_name(model: str,
     """Dataclass which contains the decoding strategy of the engine"""
 
+    DEFAULT_BACKEND: str = "auto" if envs.VLLM_USE_V1 else "xgrammar"
+
     # Which guided decoding algo to use.
     # 'outlines' / 'lm-format-enforcer' / 'xgrammar'
-    guided_decoding_backend: str = 'xgrammar'
+    guided_decoding_backend: str = DEFAULT_BACKEND
```
Is it necessary to add the `DEFAULT_BACKEND` attribute to the dataclass? Can we get away with `guided_decoding_backend: str = "auto" if envs.VLLM_USE_V1 else "xgrammar"`?

russellb replied:
I changed it to be closer to your config PR

hmellor commented on vllm/config.py:
```diff
@@ -2886,9 +2886,11 @@ def get_served_model_name(model: str,
+    DEFAULT_BACKEND: str = "auto" if envs.VLLM_USE_V1 else "xgrammar"
+
     # Which guided decoding algo to use.
     # 'outlines' / 'lm-format-enforcer' / 'xgrammar'
-    guided_decoding_backend: str = 'xgrammar'
+    guided_decoding_backend: str = DEFAULT_BACKEND
```
Oh and instead of the comment you could type hint it with `Literal["auto", "outlines", "lm-format-enforcer", "xgrammar"]`. It makes the line longer, but it's very IDE friendly.

russellb replied:
It's tricky because the valid set of values is different between V0 and V1.

hmellor commented:
I've made https://github.com/vllm-project/vllm/pull/16332 which begins to address our config code. It only covers `ParallelConfig` for now, but once we settle on a design, it can easily be extended to the other config classes.

hmellor reviewed (approved):
LGTM!

russellb committed:
Fix tests to be more V1 friendly

These tests now run against V1 in CI, but the code was originally
written with V0 in mind. In particular, specifying the backend
through the OpenAI API is not supported with V1, so we can remove
it and speed up the tests quite a bit. Different backends are tested in
another place (via the llm entrypoint tests).

Signed-off-by: Russell Bryant <rbryant@redhat.com>

alesalloum commented:
Hey, this broke my offline inference scripts with _structured outputs_ which worked perfectly in the last release 0.8.3. I tried to switch the backend to 'auto', but I still get the same error. How should one adapt their script for this change? Cheers.

`ValueError: Request-level structured output backend must match engine-level backend. xgrammar != auto`

hmellor commented:
That error indicates that the engine wants to use `auto` (the default set in this PR) but your calls to generate are setting it to `xgrammar`.

Can you try not setting the `backend` in `LLM.generate`?

If this doesn't work, please provide a minimal script so that I can try and reproduce the behaviour.

hmellor commented:
The error @alesalloum sees does not come from this PR, it was added in https://github.com/vllm-project/vllm/pull/14694. In that PR the decision was made that users cannot use arbitrary backends at runtime.

**At startup** they either leave it as `auto` and let vLLM decide which backend to use, or they pass a non-`auto` value and vLLM will always use that backend. If the user tries to set the backend at runtime they will see the error added in the PR linked above.

DarkLight1337 commented:
Oh I see. Thanks for the explanation. I think we should have a clearer error message to notify users that they are not supposed to set the backend, and add a deprecation warning for those who are currently setting the backend (even if it's the same as the one on startup).

hmellor commented:
Ok, I can make a PR for that.

Should we lower the raise to a warn and ignore with more explanation?

hmellor commented:
Here's the PR https://github.com/vllm-project/vllm/pull/16717
